lib/sync/fetcher.go

```diff
@@ -0,0 +1,256 @@
+package sync
+
+import (
+	"bufio"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"net/url"
+	"time"
+
+	"boscoin.io/sebak/lib/block"
+	"boscoin.io/sebak/lib/network"
+	"boscoin.io/sebak/lib/node"
+	"boscoin.io/sebak/lib/node/runner"
+	"boscoin.io/sebak/lib/storage"
+	"boscoin.io/sebak/lib/transaction"
+	"github.com/inconshreveable/log15"
+)
+
+type BlockFetcher struct {
+	network           network.Network
+	connectionManager network.ConnectionManager
+	apiClient         Doer
+	storage           *storage.LevelDBBackend
+	localNode         *node.LocalNode
+
+	fetchTimeout  time.Duration
+	retryInterval time.Duration
+
+	logger log15.Logger
+}
+
+type BlockFetcherOption = func(f *BlockFetcher)
+
+func NewBlockFetcher(nw network.Network,
+	cManager network.ConnectionManager,
+	st *storage.LevelDBBackend,
+	localNode *node.LocalNode,
+	opts ...BlockFetcherOption) *BlockFetcher {
+
+	f := &BlockFetcher{
+		network:           nw,
+		connectionManager: cManager,
+		apiClient:         &http.Client{},
+		storage:           st,
+		localNode:         localNode,
+		logger:            NopLogger(),
+
+		fetchTimeout:  1 * time.Minute,
+		retryInterval: 30 * time.Second,
+	}
+
+	for _, opt := range opts {
+		opt(f)
+	}
+
+	return f
+}
+
+func (f *BlockFetcher) Fetch(ctx context.Context, syncInfo *SyncInfo) (*SyncInfo, error) {
+	height := syncInfo.BlockHeight
+
+	TryForever(func(attempt int) (bool, error) {
+		select {
+		case <-ctx.Done():
+			return false, ctx.Err()
+		default:
+			f.logger.Debug("Try to fetch", "height", height, "attempt", attempt)
+			if err := f.fetch(ctx, syncInfo); err != nil {
+				if err == context.Canceled {
+					return false, ctx.Err()
+				}
+
+				f.logger.Error(err.Error(), "err", err)
+				c := time.After(f.retryInterval) //afterFunc?
+				select {
+				case <-ctx.Done():
+					return false, ctx.Err()
+				case <-c:
+					return true, err
+				}
+			}
+			return false, nil
+		}
+	})
+
+	return syncInfo, nil
+}
+
+func (f *BlockFetcher) fetch(ctx context.Context, si *SyncInfo) error {
+	var (
+		height    = si.BlockHeight
+		nodeAddrs = si.NodeAddrs
+	)
+	f.logger.Debug("Fetch start", "height", height)
+
+	n := f.pickRandomNode(nodeAddrs)
+	if n == nil {
+		return errors.New("Fetch: node not found")
+	}
+	f.logger.Info(fmt.Sprintf("fetching items from node: %v", n), "fetching_node", n, "height", height)
+
+	apiURL := apiClientURL(n, height)
+	f.logger.Debug("apiClient", "url", apiURL.String())
+
+	req, err := http.NewRequest("GET", apiURL.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancelF := context.WithTimeout(ctx, f.fetchTimeout)
+	defer cancelF()
+
+	resp, err := f.apiClient.Do(req)
+	defer func() {
+		if resp != nil && resp.Body != nil {
+			resp.Body.Close()
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode == http.StatusNotFound {
+		//TODO:
+		err := errors.New("fetch: block not found")
+		return err
```

Comment:
That's one statement to much, `return` directly

Reply:
@Geod24 I don’t understand

Reply:
```
return errors.New("fetch: block not found")
```